File: tests/Staford.NLP.Tools/Files.fs
[<AutoOpen>]
module Stanford.NLP.Tools.Files

open System.IO

[<AutoOpen>]
module Common =
    let getFullPath path =
        if File.Exists path then FileInfo(path).FullName
        elif Directory.Exists path then DirectoryInfo(path).FullName
        else failwithf "File or directory %s does not exist" path
    let combine paths = Path.Combine(Array.ofList paths) |> getFullPath

    let nlpStanford = combine [__SOURCE_DIRECTORY__; "../../data/paket-files/nlp.stanford.edu/" ]
    let dataFile path = combine [__SOURCE_DIRECTORY__; "../data/"; path]

module CoreNLP =
    let jarRoot = combine [nlpStanford; "stanford-corenlp-4.0.0/models/"]
    let models path = combine [jarRoot; "edu/stanford/nlp/models/"; path]

module NER =
    let classifier path = combine [nlpStanford; "stanford-ner-4.0.0/classifiers/"; path]

module Parser =
    let models path = combine [nlpStanford; "stanford-parser-4.0.0/models/edu/stanford/nlp/models/"; path]

module Tagger =
    let model path = combine [nlpStanford; "stanford-tagger-4.0.0/models/"; path]

module Segmenter =
    let root = combine [nlpStanford; "stanford-segmenter-4.0.0/data/"]
    let data path = combine [root; path]

File: tests/Stanford.NLP.Tools/Files.cs
using System.IO;
using System.Reflection;

namespace Stanford.NLP.Tools
{
    public static class Files
    {
        private static string? _rootFolder;
        private static string GetRootFolder()
        {
            if (_rootFolder is not null)
                return _rootFolder;

            _rootFolder = new FileInfo(Assembly.GetExecutingAssembly().Location).DirectoryName;
            while (true)
            {
                if (File.Exists(Path.Combine(_rootFolder, ".gitignore")))
                    return _rootFolder;
                _rootFolder = new DirectoryInfo(_rootFolder!).Parent.FullName;
            }
        }

        private static string NlpStanford =>
            Path.Combine(GetRootFolder(), "data/paket-files/nlp.stanford.edu/");

        public static string DataFile(string path) =>
            Path.Combine(GetRootFolder(), "tests/data/", path);

        public static class CoreNlp
        {
            public static string JarRoot =>
                Path.Combine(NlpStanford, "stanford-corenlp-4.3.2/models/");

            public static string Models(string path) =>
                Path.Combine(JarRoot, "edu/stanford/nlp/models/", path);
        }

        public static class NER
        {
            public static string Classifier(string path) =>
                Path.Combine(NlpStanford, "stanford-ner-2020-11-17/classifiers/", path);
        }

        public static class Parser
        {
            public static string Models(string path) =>
                Path.Combine(NlpStanford, "stanford-parser-full-2020-11-17/models/edu/stanford/nlp/models/", path);
        }

        public static class Tagger
        {
            public static string Model(string path) =>
                Path.Combine(NlpStanford, "stanford-postagger-full-2020-11-17/models/", path);
        }

        public static class Segmenter
        {
            public static string Root =>
                Path.Combine(NlpStanford, "stanford-segmenter-2020-11-17/data/");

            public static string Data(string path) =>
                Path.Combine(Root, path);
        }
    }
}
